enzyme/Enzyme/EnzymeLogic.cpp

```diff
@@ -4236,6 +4237,256 @@ Function *EnzymeLogic::CreateForwardDiff(
+
+  ValueToValueMapTy originalToNewFn;
+
+  // Create placeholder for the old arguments
+  BasicBlock *placeholderBB =
+      BasicBlock::Create(NewF->getContext(), "placeholders", NewF);
+
+  IRBuilder<> PlaceholderBuilder(placeholderBB);
+  PlaceholderBuilder.SetCurrentDebugLocation(DebugLoc());
+  ValueToValueMapTy vmap;
+  auto DestArg = NewF->arg_begin();
+
+  for (unsigned i = 0; i < orig_FTy->getNumParams(); ++i) {
+    Argument *arg = tobatch->getArg(i);
+    if (arg_types[i] == BATCH_TYPE::VECTOR) {
+      auto placeholder = PlaceholderBuilder.CreatePHI(
+          arg->getType(), 0, "placeholder." + arg->getName());
+      vmap[arg] = placeholder;
+      DestArg->setName(arg->getName());
+      DestArg++;
+    } else {
+      vmap[arg] = NewF->getArg(i);
+    }
+  }
+
+  SmallVector<ReturnInst *, 4> Returns;
+  CloneFunctionInto(NewF, tobatch, vmap, true, Returns);
+
+  NewF->setLinkage(Function::LinkageTypes::InternalLinkage);
+
+  // find instructions to vectorize (going up / overestimation)
+  SmallPtrSet<Value *, 32> toVectorize;
+  SetVector<llvm::Value *, std::deque<llvm::Value *>> worklist;
+
+  for (unsigned i = 0; i < tobatch->getFunctionType()->getNumParams(); i++) {
+    if (arg_types[i] == BATCH_TYPE::VECTOR) {
+      Argument *arg = tobatch->arg_begin() + i;
+      toVectorize.insert(arg);
+
+      for (auto user : arg->users()) {
+        worklist.insert(user);
+      }
+    }
+  }
+
+  if (!NewF->getReturnType()->isVoidTy())
+    for (auto &BB : *tobatch)
+      if (auto ret = dyn_cast<ReturnInst>(BB.getTerminator()))
+        worklist.insert(ret);
+
+  while (!worklist.empty()) {
+    Value *todo = *worklist.begin();
+    worklist.erase(worklist.begin());
+
+    if (isa<ConstantData>(todo))
+      continue;
+
+    if (isa<Argument>(todo))
+      continue;
+
+    if (toVectorize.count(todo) != 0) {
+      continue;
+    }
+
+    toVectorize.insert(todo);
+
+    if (auto inst = dyn_cast<Instruction>(todo)) {
+      for (auto user : inst->users()) {
+        worklist.insert(user);
+      }
+      for (auto &user : inst->operands()) {
+        worklist.insert(user);
+      }
+    }
+  }
+
+  // find instructions to vectorize (going down / refine)
+  SetVector<llvm::Value *, std::deque<llvm::Value *>> refinelist;
+
+  for (unsigned i = 0; i < tobatch->getFunctionType()->getNumParams(); i++) {
+    if (arg_types[i] == BATCH_TYPE::SCALAR) {
+      Argument *arg = tobatch->arg_begin() + i;
+
+      for (auto user : arg->users()) {
+        refinelist.insert(user);
+      }
+    }
+  }
+
+  while (!refinelist.empty()) {
+    Value *todo = *refinelist.begin();
+    refinelist.erase(refinelist.begin());
+
+    if (isa<ReturnInst>(todo))
+      continue;
+
+    if (isa<CallInst>(todo))
+      continue;
+
+    if (auto todo_inst = dyn_cast<Instruction>(todo)) {
+
+      if (todo_inst->mayReadOrWriteMemory())
+        continue;
+
+      bool br = false;
+      for (auto &arg : todo_inst->operands()) {
```

Comment:
Let's also have this work on phi nodes.

The algorithm for this would be as follows:
toCheck = todo_inst->operands();
safe = set();
legal = true;
while (!toCheck.empty()) {
  cur = todo.pop_front();
  if (safe.count(cur)) continue;
  safe.insert(cur);
  if (cur is scalar) continue;
  if (cur is not call and doesn't write memory) {
    add all cur's operands.
    continue
  }
  legal = false; break;
}